design_drafts/async_controller.md

```diff
@@ -0,0 +1,112 @@
+# Asynchronous Controllers in ros2_control
+
+## What is an asynchronous controller?
+
+An asynchronous controller is a controller that for some reason cannot (or we don’t want to) perform the operations needed in an update() call.
+For instance if ros control is running at 100Hz, the sum of the execution time of all controllers update() calls must be below 10ms. If a controller requires 15ms it cannot be executed synchronously without affecting the whole ros control update rate.
+
+## Abstract implementation description
+Create an AsyncControllerWrapper class that can wrap any Controller without modifying it and makes it run asynchronously seamlessly.
+It will receive a Controller and store it, and act as an intermediary between the ros_control loop and the Controller.
+On the update() call, the AsyncControllerWrapper will read data from the interfaces, store it in an intermediate object, and write data from another intermediate object to the interfaces.
+The wrapped Controller, running on a separate thread, will read and write from these intermediate objects.
+This data must be protected by mutexes or similar.
+When new state data is available, the AsyncControllerWrapper will notify the wrapped controller so it is processed on its own thread.
+
+## Implementation guidelines:
+
+### Concurrency risks
+The main risk is handling concurrency, since we want to wrap any controller, which was probably not implemented to support concurrent calls to some of its methods.
+The only method of the wrapped controller that we’ll run in a separate thread, is update().
+The rest of the method calls will be forwarded from the AsyncControllerWrapper to the wrapped controller in the main thread.
+We must make sure that the asynchronous update() is not being executed when forwarding other methods.
+
+### StateInterfaces and CommandInterfaces
+The second risk we want to control, is that our asynchronous controller has access to State and Command Interfaces, and uses them from its own thread, causing for instance concurrent write of a CommandInterface while the Actuator is reading it.
+
+To avoid this, we can create intermediate Interfaces, stored in the AsyncControllerWrapper, that act as a bridge between the wrapped controller and the real State and Command Interfaces.
+This bridge should allow thread safe data storage, and non blocking operations on the ros_control thread. Some containers for this purpose exist in: https://github.com/ros-controls/realtime_tools/tree/ros2_devel/include/realtime_tools
+
+These interfaces should be created on the assign_interfaces method, and removed on the release_interfaces method. But at the moment those methods are not virtual.
+I’ll assume they will be changed to virtual, but if that cannot be done for ABI compatibility issues, the same could be done in the on_activate and on_deactivate, although it might be less correct.
```

Comment:
I don't really count this as an argument. We should do this in a “correct” way. Regardless of ABI compatibility.

Reply:
had to be listed for correctness ;)

Reply:
```suggestion
These interfaces should be created on the assign_interfaces method, and removed on the release_interfaces method. But at the moment those methods are not virtual, that will need to change.
```